WIP: GSoC Git Support - Support semantic pull merge without conflicts

wanling0000 opened the pull request:
Closes #12350

### Steps to test

This PR implements the first part of Git sync support by enabling automatic semantic merges when there are no conflicts. The scenario is tested via TDD in GitSyncServiceTest, simulating the following steps:

1. Alice creates an initial commit and pushes it to the remote.
2. Bob clones the repo and modifies entry b, then pushes.
3. Alice modifies entry a, then pulls.

Since they changed different entries, we expect a clean merge without user intervention. The orchestrator is `GitSyncService`, and helper utilities and value objects are temporarily located in `org.jabref.logic.git.util`.

### Mandatory checks

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [.] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [.] Screenshots added in PR description (if change is visible to the user)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [.] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/GitFileWriter.java:
```diff
@@ -0,0 +1,40 @@
+import org.jabref.model.entry.BibEntryTypesManager;
+
+public class GitFileWriter {
+    public static void write(Path file, BibDatabaseContext ctx, ImportFormatPreferences importPrefs) throws IOException {
```
We generally avoid abbreviations (that would actually be an understatement - as we ae Java people and thus prefer verbose names :P), so please change `ctx` to `context` or `databaseContext`.

Also fun fact: "ctx" is more commonly used in context of parse tree nodes in compilers/parsers (ANTLR etc.).

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/GitRevisionLocator.java:
```diff
@@ -0,0 +1,38 @@
+ * remote = origin/main
+ */
+public class GitRevisionLocator {
+    public RevisionTriple locateMergeCommits(Git git) throws Exception {
```
```suggestion
    public RevisionTriple locateMergeCommits(Git git) throws IOException {
```

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/GitRevisionLocator.java:
```diff
@@ -0,0 +1,38 @@
+public class GitRevisionLocator {
+    public RevisionTriple locateMergeCommits(Git git) throws Exception {
+        // assumes the remote branch is 'origin/main'
+        ObjectId headId = git.getRepository().resolve("HEAD");
```
Extract the string into a constant `HEAD`.

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/GitRevisionLocator.java:
```diff
@@ -0,0 +1,38 @@
+        ObjectId headId = git.getRepository().resolve("HEAD");
+        // and uses the default remote tracking reference
+        // does not support multiple remotes or custom remote branch names so far
+        ObjectId remoteId = git.getRepository().resolve("refs/remotes/origin/main");
```
Likewise, extract the string to a constant `REMOTE`.

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/MergeResult.java:
```diff
@@ -0,0 +1,19 @@
+
+public record MergeResult(boolean successful, List<BibEntryDiff> conflicts) {
+    public static MergeResult conflictsFound(List<BibEntryDiff> conflicts) {
+        return new MergeResult(false, conflicts);
```
Extract to `SUCCESS = true` and use `!SUCCESS`

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/MergeResult.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    public static MergeResult success() {
+        return new MergeResult(true, List.of());
```
`SUCCESS`

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/RevisionTriple.java:
```diff
@@ -0,0 +1,15 @@
+ * @param local the current local branch tip
+ * @param remote the tip of the remote tracking branch (typically origin/main)
+ */
+public record RevisionTriple(
+        RevCommit base, RevCommit local, RevCommit remote) { }
```
Nitpick: Since it's one line of parameters anyway
```suggestion
public record RevisionTriple(RevCommit base, RevCommit local, RevCommit remote) { }
```

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/SemanticConflictDetector.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    public static Map<String, BibEntry> toEntryMap(BibDatabaseContext context) {
+        return context.getDatabase().getEntries().stream()
+                      .filter(e -> e.getCitationKey().isPresent())
+                      .collect(Collectors.toMap(
+                              e -> e.getCitationKey().get(),
+                              Function.identity(),
+                              (a, b) -> b,
+                              LinkedHashMap::new
+                      ));
```
```suggestion
        return context.getDatabase().getEntries().stream()
                      .filter(entry -> entry.getCitationKey().isPresent())
                      .collect(Collectors.toMap(
                              entry -> entry.getCitationKey().get(),
                              Function.identity(),
                              (existing, replacement) -> replacement,
                              LinkedHashMap::new
                      ));
```

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/SemanticConflictDetector.java:
```diff
@@ -0,0 +1,142 @@
+            boolean remoteChanged = !Objects.equals(baseVal, remoteVal);
+
+            if (localChanged && remoteChanged && !Objects.equals(localVal, remoteVal)) {
+                return true;
+            }
+        }
+
+        return false;
```
Since the method is short enough, magic booleans are okay here 👍🏻

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/SemanticConflictDetector.java:
```diff
@@ -0,0 +1,142 @@
+        return false;
+    }
+
+    public static MergePlan extractMergePlan(BibDatabaseContext base, BibDatabaseContext remote) {
```
This method could make use of some javadoc.

subhramit commented on jablib/src/main/java/org/jabref/logic/git/util/SemanticConflictDetector.java:
```diff
@@ -0,0 +1,142 @@
+        // 1. get diffs between base and remote
+        List<BibEntryDiff> remoteDiffs = BibDatabaseDiff.compare(base, remote).getEntryDifferences();
+        if (remoteDiffs == null) {
+            return Collections.emptyList();
```
```suggestion
            return List.of();
```

subhramit commented on jablib/src/test/java/org/jabref/logic/git/util/GitFileWriterTest.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Test
+    void testWriteThenReadBack() throws Exception {
+        BibDatabaseContext inputCtx = GitBibParser.parseBibFromGit(
```
`inputDatabaseContext`

subhramit commented on jablib/src/test/java/org/jabref/logic/git/util/GitBibParserTest.java:
```diff
@@ -0,0 +1,77 @@
+        List<BibEntry> entries = context.getEntries();
+        assertEquals(1, entries.size());
+
+        BibEntry entry = entries.get(0);
```
Last few are just idiomatic nitpicks
```suggestion
        BibEntry entry = entries.getFirst();
```

subhramit commented on jablib/src/test/java/org/jabref/logic/git/util/GitFileWriterTest.java:
```diff
@@ -0,0 +1,51 @@
+        List<BibEntry> outputEntries = outputCtx.getDatabase().getEntries();
+
+        assertEquals(inputEntries.size(), outputEntries.size());
+        assertEquals(inputEntries.get(0).getCitationKey(), outputEntries.get(0).getCitationKey());
+        assertEquals(inputEntries.get(0).getField(StandardField.AUTHOR), outputEntries.get(0).getField(StandardField.AUTHOR));
```
```suggestion
        assertEquals(inputEntries.getFirst().getCitationKey(), outputEntries.getFirst().getCitationKey());
        assertEquals(inputEntries.getFirst().getField(StandardField.AUTHOR), outputEntries.getFirst().getField(StandardField.AUTHOR));
```

subhramit commented on jablib/src/test/java/org/jabref/logic/git/util/SemanticConflictDetectorTest.java:
```diff
@@ -0,0 +1,483 @@
+
+        RevCommit baseCommit = writeAndCommit(base, "base", alice);
+        RevCommit remoteCommit = writeAndCommit(remote, "remote", bob);
+        BibDatabaseContext baseCtx = parse(baseCommit);
+        BibDatabaseContext remoteCtx = parse(remoteCommit);
```
baseDatabaseContext, remoteDatabaseContext

subhramit commented on jablib/src/test/java/org/jabref/logic/git/util/SemanticMergerTest.java:
```diff
@@ -0,0 +1,110 @@
+    @ParameterizedTest(name = "Database patch: {0}")
+    @MethodSource("provideDatabasePatchCases")
+    void testPatchDatabase(String description, String base, String local, String remote, String expectedAuthor) throws Exception {
+        BibDatabaseContext baseCtx = GitBibParser.parseBibFromGit(base, importFormatPreferences);
+        BibDatabaseContext localCtx = GitBibParser.parseBibFromGit(local, importFormatPreferences);
+        BibDatabaseContext remoteCtx = GitBibParser.parseBibFromGit(remote, importFormatPreferences);
+
+        MergePlan plan = SemanticConflictDetector.extractMergePlan(baseCtx, remoteCtx);
+        SemanticMerger.applyMergePlan(localCtx, plan);
+
+        BibEntry patched = localCtx.getDatabase().getEntryByCitationKey("a").orElseThrow();
```
All Ctx -> DatabaseContext

subhramit commented on jablib/src/test/java/org/jabref/logic/git/util/SemanticConflictDetectorTest.java:
```diff
@@ -0,0 +1,483 @@
+        RevCommit localCommit = writeAndCommit(local, "local", alice);
+        RevCommit remoteCommit = writeAndCommit(remote, "remote", bob);
+
+        BibDatabaseContext baseCtx = parse(baseCommit);
+        BibDatabaseContext localCtx = parse(localCommit);
+        BibDatabaseContext remoteCtx = parse(remoteCommit);
```
all Ctx -> DatabaseContext

subhramit reviewed (changes requested):
This was not easy - really beautiful work, Wanling. For a WIP, this is exceptionally good!
I am leaving some comments - the earlier we get habituated to some idiomatic practices (some JabRef-specific, some general), the easier it is to stick to them during the rest of the project.

Please don't be overwhelmed by the number of comments - most of them are minor. Your code is already good, so the only way I could help make it better was by nit-picking.

koppor reviewed (commented):
Small comment on the process

wanling0000 commented on jablib/src/main/java/org/jabref/logic/git/GitSyncService.java:
```diff
@@ -0,0 +1,128 @@
+        MergeResult result = performSemanticMerge(git, triple.base(), triple.local(), triple.remote(), bibFilePath);
+
+        // 4. Automatic merge
+        if (result.successful()) {
```
Thanks! I hadn't thought about how much smoother that reads. Naming really is an art 😄

wanling0000 commented on jablib/src/main/java/org/jabref/logic/git/GitSyncService.java:
```diff
@@ -0,0 +1,128 @@
+        BibDatabaseContext remote = GitBibParser.parseBibFromGit(remoteContent, importFormatPreferences);
+
+        // 2. Conflict detection
+        List<BibEntryDiff> conflicts = SemanticConflictDetector.detectConflicts(base, local, remote);
+
+        if (!conflicts.isEmpty()) {
+            // Currently only handles non-conflicting cases. In the future, it may:
+            // - Store the current state along with 3 versions
+            // - Return conflicts along with base/local/remote versions for each entry
+            // - Invoke a UI merger (let the UI handle merging and return the result)
+            return MergeResult.conflictsFound(conflicts); // Conflicts: return the conflict result and let the UI layer handle it
```
Thanks! I’ve renamed it to MergeResult.withConflicts(conflicts) for now, and I left a TODO here to revisit the naming as the implementation evolves

InAnYan commented on jablib/src/main/java/org/jabref/logic/git/GitSyncService.java:
```diff
@@ -0,0 +1,128 @@
+public class GitSyncService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GitSyncService.class);
+    private final ImportFormatPreferences importFormatPreferences;
+    private GitHandler gitHandler;
```
git handler could be changed? if not, then it's better to add `final`

InAnYan commented on jablib/src/main/java/org/jabref/logic/git/GitSyncService.java:
```diff
@@ -0,0 +1,128 @@
+        // 4. Automatic merge
+        if (result.successful()) {
+            gitHandler.createCommitOnCurrentBranch("Auto-merged by JabRef", false);
+        }
```
Mhm, I see. If there is a conflict, then it would not change anything, but if there are no conflicts, then the result will be successful

InAnYan commented on jablib/src/main/java/org/jabref/logic/git/util/SemanticMerger.java:
```diff
@@ -0,0 +1,61 @@
+
+    /**
+     * Implementation-only merge logic: applies changes from remote (relative to base) to local.
+     * does not check for "modifications" or "conflicts" — all decisions should be handled in advance by the SemanticConflictDetector
```
Small moment:
```suggestion
     * does not check for "modifications" or "conflicts" — all decisions should be handled in advance by the {@link SemanticConflictDetector}
```

wanling0000 commented on jablib/src/main/java/org/jabref/logic/git/GitSyncService.java:
```diff
@@ -0,0 +1,128 @@
+        if (bibPath.startsWith(workTree)) {
+            relativePath = workTree.relativize(bibPath);
+        } else {
+            throw new IllegalStateException("Given .bib file is not inside repository");
```
Added a TODO at this point because the orchestration logic (e.g., GitPullViewModel or GitSyncService) has been restructured. I agree it's better to validate the .bib path earlier

wanling0000 commented:
Sorry, I didn’t realize this branch had been rebased earlier. After I merged main, the commit history became quite messy. Try to fix it now 😭

koppor commented:
> Sorry, I didn’t realize this branch had been rebased earlier. After I merged main, the commit history became quite messy. Try to fix it now 😭

Not sure if its really a rebase. The comments are still shown.

Use "gitk --all" to investigate.

We can also do together...

wanling0000 commented:
> Not sure if its really a rebase. The comments are still shown.
> 
> Use "gitk --all" to investigate.
> 
> We can also do together...

Sorry I didn’t explain clearly earlier!

At the very beginning, I rebased my branch onto main. But later, when I tried to push again, I accidentally merged upstream/main into it, which caused all the merged commits from main to appear in my history 🙈

I’ve just cleaned it up using cherry-pick onto a fresh branch based on main, and force-pushed it to keep the commit history clean. Hopefully it looks better now...

(Also, please ignore the gui module in this PR for now — as discussed earlier, I’ll open a separate PR for the GUI-related updates)
